src/main/java/com/yelp/nrtsearch/server/luceneserver/SearchHandler.java

```diff
@@ -191,27 +190,54 @@ public SearchResponse handle(IndexState indexState, SearchRequest searchRequest)
       hitBuilders.sort(Comparator.comparing(Hit.Builder::getLuceneDocId));
       List<LeafReaderContext> leaves = s.searcher.getIndexReader().leaves();
 
+      List<LeafReaderContext> hitIdToLeaves = new ArrayList<>();
       for (int hitIndex = 0; hitIndex < hitBuilders.size(); ++hitIndex) {
         var hitResponse = hitBuilders.get(hitIndex);
-
         LeafReaderContext leaf =
             leaves.get(ReaderUtil.subIndex(hitResponse.getLuceneDocId(), leaves));
+        hitIdToLeaves.add(hitIndex, leaf);
+      }
 
-        if (!searchContext.getRetrieveFields().isEmpty()) {
-          var fieldValueMap =
-              fillFields(
+      for (String field : searchContext.getRetrieveFields().keySet()) {
+        for (int hitIndex = 0; hitIndex < hitBuilders.size(); ++hitIndex) {
+          var hitResponse = hitBuilders.get(hitIndex);
+          LeafReaderContext leaf = hitIdToLeaves.get(hitIndex);
+          CompositeFieldValue v =
+              getFieldForHit(
```

Comment:
I guess you can try this approach to see if it gives you enough perf improvement. This could be made better by the reuse of per segment resources.

Reply:
I am not sure how to reuse the segment here.
The doc value field is column based. So it is recommended to iterate by fields then id. but segment is by id. I checked leafreader source code. There are a ton of implementations and leafreader instance in leafreader classes. not sure if that will help

Reply:
The idea would be to create the per segment object (ScoreScript, LoadedDocValues<>, etc) once, then advance the doc id for subsequent segment documents.

If you want, you can try your current approach first, since it is a more straightforward change.

Reply:
ack, let's try this first. I'll have follow-up prs since this optimization actually make stored fields slower since that is row based. I'll have follow-up prs to treat them differently